### PR TITLE
Install button added to the UI

### PR DESCRIPTION
**Pull request by jvelezpo**
* Prevents opening many instances of the app
* Install button added to the UI (without functionality)
* differentiate between installed and not installed editors 
* use redux action
* only show installed editors
* use stateless components

### CODE REVIEW

File: src/components/ActiveEditors.tsx [hunk 1]
```diff
@@ -3,13 +3,18 @@ import PropTypes from "prop-types";
...
 
-const ActiveEditors = ({ editors }) => (
-  <div>
-    {editors.map(editor => (
-      <EditorIcon {...editor} />
-    ))}
-  </div>
-);
+const ActiveEditors = ({ editors }) => {
+  // console.log('=============================');
```
Comment (alanhamlett) [inline src/components/ActiveEditors.tsx]: We can remove this commented out code.

File: src/constants/editors.ts [hunk 2]
```diff
@@ -9,12 +18,12 @@ export const VSCODE = "VSCODE";
...
 export default [
-  ATOM,
-  CHROME,
-  ECLIPSE,
-  ANDROIDSTUDIO,
-  SUBLIMETEXT2,
-  SUBLIMETEXT3,
-  VIM,
-  VSCODE
+  { name: ATOM, instance: new Atom() },
```
Comment (alanhamlett) [inline src/constants/editors.ts]: Each editor class has it's own `name` attribute. Let's just `export default [Atom, SublimeText2...` then instantiate each class in the local scope where they're used.

File: src/constants/editors.ts [hunk 3]
```diff
@@ -1,3 +1,12 @@
+import Atom from "../editors/atom";
+import SublimeText2 from "../editors/sublime-text-2";
+import SublimeText3 from "../editors/sublime-text-3";
+import Vim from "../editors/vim";
+import Chrome from "../editors/chrome";
+import Eclipse from "../editors/eclipse";
+import AndroidStudio from "../editors/android-studio";
+import VSCode from "../editors/vscode";
+
 export const ANDROIDSTUDIO = "ANDROIDSTUDIO";
```
Comment (alanhamlett) [inline src/constants/editors.ts]: These were here before your changes, but let's remove them and use editor classes directly.

File: src/components/EditorIcon.tsx [hunk 4]
```diff
@@ -1,36 +1,54 @@
...
-    <div {...css(styles.wrapper)}>
-      <img
-        {...css(styles.img, enabled ? styles.enabled : styles.disabled)}
-        alt={name}
-        src={img}
-      />
-    </div>
-  );
-};
+interface IProps {
```
Comment (alanhamlett) [inline src/components/EditorIcon.tsx]: Does `Props` work or is there a name collision?

File: src/components/EditorIcon.tsx [hunk 5]
```diff
@@ -1,36 +1,54 @@
...
+  img: string;
+  name: string;
+  height: number | string;
+  width: number | string;
+  maxHeight: number | string;
+  styles: IStyles;
+  instance: EditorInterface;
+  css: Function;
+}
+interface IState {
```
Comment (alanhamlett) [inline src/components/EditorIcon.tsx]: `State`?

File: src/components/EditorIcon.tsx [hunk 6]
```diff
@@ -1,36 +1,54 @@
...
+  enabled: boolean;
+}
+interface IStyles {
```
Comment (alanhamlett) [inline src/components/EditorIcon.tsx]: Same, `Styles`.

File: src/components/EditorIcon.tsx [hunk 7]
```diff
@@ -1,36 +1,54 @@
...
+  wrapper: Object;
+  img: Object;
+  enabled: Object;
+  disabled: Object;
+}
+class EditorIcon extends React.Component<IProps, IState> {
```
Comment (alanhamlett) [inline src/components/EditorIcon.tsx]: Let's use `PureComponent` wherever we can, so this component should be stateless.

File: src/components/EditorIcon.tsx [hunk 8]
```diff
@@ -1,36 +1,54 @@
...
+
+  constructor(props) {
+    super(props);
+    this.state = {
+      enabled: false
+    };
+  }
+  public async componentDidMount(): Promise<void> {
+    const { instance } = this.props;
+    this.setState({ enabled: await instance.isEditorInstalled() });
```
Comment (alanhamlett) [inline src/components/EditorIcon.tsx]: This is nice, but it spreads logic around the app. I prefer to call `isEditorInstalled()` in parent and keep this class stateless for easier unit testing and improved performance.

File: src/editors/editor.ts [hunk 9]
```diff
@@ -11,8 +11,12 @@ export default abstract class Editor implements EditorInterface {
   }
 
   public async isDirectory(directory: string): Promise<boolean> {
-    const stats = await fs.stat(directory);
-    return stats.isDirectory();
+    try {
+      const stats = await fs.stat(directory);
+      return stats.isDirectory();
+    } catch (err) {
+      return false;
```
Comment (alanhamlett) [inline src/editors/editor.ts]: Thanks!

File: src/reducers/editors.ts [hunk 10]
```diff
@@ -5,9 +5,10 @@ import createReducer from "../utils/createReducer";
 const initialState = Object.values(editors).reduce((accum, val) => {
   // eslint-disable-next-line no-param-reassign
   accum.push({
-    name: val,
+    name: val.name,
```
Comment (alanhamlett) [inline src/reducers/editors.ts]: If we use editor classes directly, the base editor class should have an `id` or `key` method that creates a unique slug from `instance.name`.

File: src/reducers/editors.ts [hunk 11]
```diff
@@ -5,9 +5,10 @@ import createReducer from "../utils/createReducer";
...
     enabled: true, // Enabled by default for now
-    img: imgPathMap[val]
+    img: imgPathMap[val.name],
```
Comment (alanhamlett) [inline src/reducers/editors.ts]: If we use editor classes directly, the base editor class should have an `id` or `key` method that creates a unique slug from `instance.name`.

File: src/constants/editors.ts [hunk 12]
```diff
@@ -6,24 +6,16 @@ import Chrome from "../editors/chrome";
...
+  Atom,
+  Chrome,
+  Eclipse,
+  AndroidStudio,
+  SublimeText2,
+  SublimeText3,
+  Vim,
+  VSCode,
+  XCode
+};
```
Comment (gandarez) [inline src/constants/editors.ts]: There're more editors ready to be added into thsi list. Could you please add them?

Comment (jvelezpo) [reply src/constants/editors.ts]: yes, let me add all of them 👍

Comment (jvelezpo) [reply src/constants/editors.ts]: done

**Pull request merged**